memstore/kafka_redolog_manager.go

```diff
@@ -0,0 +1,168 @@
+func (k *kafkaRedologManager) WriteUpsertBatch(upsertBatch *UpsertBatch) (int64, uint32) {
+	panic("WriteUpsertBatch to kafka redolog manager is disabled")
+}
+
+func (k *kafkaRedologManager) UpdateMaxEventTime(eventTime uint32, fileID int64) {
+	k.Lock()
+	defer k.Unlock()
+	if _, ok := k.MaxEventTimePerFile[fileID]; ok && eventTime <= k.MaxEventTimePerFile[fileID] {
+		return
+	}
+	k.MaxEventTimePerFile[fileID] = eventTime
+}
+
+// NextUpsertBatch creates iterator function for getting the next upsert batch
+func (k *kafkaRedologManager) NextUpsertBatch() func() (*UpsertBatch, int64, uint32) {
+	return func() (*UpsertBatch, int64, uint32) {
+		for {
+			select {
+			case batchBundle := <-k.batches:
+				fileID := batchBundle.kafkaOffset / maxBatchesPerFile
+				fileOffset := batchBundle.kafkaOffset % maxBatchesPerFile
+				return batchBundle.batch, fileID, uint32(fileOffset)
+			case <-k.done:
+				// redolog manager closed
+				return nil, 0, 0
+			}
+		}
+	}
+}
+
+func (k *kafkaRedologManager) PurgeRedologFileAndData(eventTimeCutoff uint32, fileIDCheckpointed int64, batchOffset uint32) error {
+	k.RLock()
+	var firstUnpurgeable int64 = math.MaxInt64
+	for fileID, maxEventTime := range k.MaxEventTimePerFile {
+		if maxEventTime >= eventTimeCutoff ||
+			fileID > fileIDCheckpointed ||
+			(fileID == fileIDCheckpointed && batchOffset != maxBatchesPerFile-1) {
+			// file not purgeable
+			if fileID < firstUnpurgeable {
+				firstUnpurgeable = fileID
+			}
+		}
+	}
+	k.RUnlock()
+
+	if firstUnpurgeable < math.MaxInt64 {
+		err := k.commitFunc(firstUnpurgeable * maxBatchesPerFile)
+		if err != nil {
+			return err
+		}
+
+		k.Lock()
+		for fileID := range k.MaxEventTimePerFile {
+			if fileID < firstUnpurgeable {
+				delete(k.MaxEventTimePerFile, fileID)
+			}
+		}
+		k.Unlock()
+	}
+	return nil
+}
+
+// ConsumeFrom starts kafka consumer from a offset
+func (k *kafkaRedologManager) ConsumeFrom(offset int64) error {
+	if k.partitionConsumer != nil {
+		// close previous created partition consumer
+		k.partitionConsumer.Close()
+	}
+	var err error
+	k.partitionConsumer, err = k.consumer.ConsumePartition(k.Topic, int32(k.Shard), offset)
+	if err != nil {
+		return utils.StackError(err, "failed to consume from topic %s partition %d",
+			k.Topic, k.Shard)
+	}
+
+	go func() {
+		for {
+			select {
+			case msg, ok := <-k.partitionConsumer.Messages():
+				if !ok {
+					// consumer closed
+					return
+				}
+				if msg != nil {
+					upsertBatch, err := NewUpsertBatch(msg.Value)
+					if err != nil {
+						utils.GetLogger().With(
+							"table", k.TableName,
+							"Shard", k.Shard).Error("failed to create upsert batch from msg")
+						continue
+					}
+					k.batches <- upsertBatchBundle{upsertBatch, msg.Offset}
+				}
+			case <-k.done:
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (k *kafkaRedologManager) Close() {
+	if k.partitionConsumer != nil {
+		k.partitionConsumer.Close()
```

Comment:
if we don't set nil here, would it panic at line 118?

Reply:
if user close redolog manager, they should not start consumer from it